Various updates

sntran opened the pull request:
## Features
- Function to retrieve channels
- Function to retrieve playlists
- Function to retrieve videos with details
- Function to search
- `playlistItems` and `playlistItemsWithTags` can retrieve `uploads` playlist ID if not set in `params`
## Performance Improvements
- `playlistItemsWithTags` now makes less requests thanks to `videos` function. `videos` endpoint can take a comma-separated list of IDs instead, so we can get details of 50 videos per request, reducing # requests from `N` to `N div 50`.
- Support passing `params.key` when OAuth is not needed.
## Bug Fixes
- Clean up for jshint ( @jlafitte )
- Uniform tab vs spaces ( @jlafitte )
- Return error in API response's body as actual error
- Ensure credentials are used. `spas-http` does not handle `access_token` in `credentials`, so we need
  to add it to the `params` manually so we don't have limit restriction.
- Fix a bug when `credentials` is not provided. When a bundle does not require authentication, credentials is null, which would cause spas to crash.
## Breaking Changes
-  `playlistItems` and `playlistItemsWithTags` both now return the full structure instead of just snippet. This is to ensure when `params.part` is set, correct data are returned to user.
- `playlistItemsWithTags` returns data with a little difference than `playlistItems`. They are the `snippet`, or `contentDetails` or any that are of `video` resource. `playlistItems` returns data of `playlistItem` resource.

jlafitte committed:
Cleaned up some things so jshint would be happy, nothing major.

jlafitte committed:
Pass through the part param to the video request.

jlafitte committed:
Uniform tab vs spaces

Commit:
Make results consistent between two functions

The data from `playlistItem` and `video` type are different. We only
need the `snippet` object in `playlistItem` and the tags from `video`.

Both functions return the following structure:

    [{
        publishedAt: true,
        channelId: true,
        title: true,
        description: true,
        thumbnails: {
          default: {
            url: true,
            width: true,
            height: true
          },
          medium: true,
          high: true,
          standard: true,
          maxres: true
        },
        channelTitle: true,
        playlistId: true,
        position: 0,
        resourceId: {
          kind: true,
          videoId: true
        },
        id: true,
        tags: true
      }
    ]

Commit:
Add `access_token` to `playlistItem.list` request

`spas-http` does not handle `access_token` in `credentials`, so we need
to add it to the `params` manually so we don't have limit restriction.

Commit:
Handle errors

YouTube does not return error status code, but rather in the response

Commit:
Support default `playlistId` 

It's not trivial to retrieve `playlistId` of the `uploads` playlist to
use in `params`. This adds support to retrieve the `uploads` playlist
ID when it is not set in `params`.

Commit:
Extract the `channels` function

And provide it as another export. It's very basic right now and mostly
to be used to get the `uploads` playlist ID, but can be extended if
needed later on.

Commit:
Revert back to return full item

Because of the support for `params.part`, it's better to return the
objects specified in there. After all, if the user wants `.status`,
they may want to filter the results, so we give them data.

Commit:
Fix a bug when `credentials` not provided

When a bundle does not require authentication, credentials is null,
which will cause spas to crash. This fixes it.

Commit:
More error handling

Commit:
Support `params.forUsername ` to filter channel

Commit:
Pass through YouTube's error

Seems like SPAS does not like `Error` instance.

Commit:
Support passing an Google API key

Using API server key will increase usage limit, without requiring
OAuth

Commit:
Reduce number of requests

`videos` endpoint can take a comma-separated list of IDs instead, so
we can get details of 50 videos per request, reducing # requests from
N to N div 50.

This commit also changes the return value of `playlistItemsWithTags`.
It now overrides with the full details of the videos, not just tags.

It is to allow user to get other properties such as video's duration,
which is under `contentDetails`.

The user can use `filter` to remove unwanted properties.

Commit:
Add function to retrieve list of playlists

Commit:
Add `search` function

Supports all parameters from
https://developers.google.com/youtube/v3/docs/search/list,
with the exception of `maxResults`. It can be set higher
than YouTube limit of 50.

Commit:
Support requesting any number of playlists

Commit:
Merge remote-tracking branch 'john/master' into patch-1

jlafitte commented:
I don't think I've ever seen a fancier commit message.

jlafitte commented:
@dtex Any chance we can get this pull merged and updated in NPM as well?  The current version breaks, I had to manually cope this over.

sntran committed:
Support statistics part

The `videos` endpoint supports `statistics` part for views count, likes, etc...
However, when using it in `playlistItems` or `playlistItemsWithTags`, those
endpoints do not take `statistics` part, so we have to remove it there.

sntran committed:
Update version
